{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "import subprocess\n",
    "\n",
    "\n",
    "def ko2genes(ko):\n",
    "    url = 'http://rest.kegg.jp/link/genes/' + ko\n",
    "    r = requests.get(url)\n",
    "    geneKO = pd.DataFrame([x.split('\\t') for x in r.text.split('\\n')],columns=['ko','gene'])\n",
    "    geneKO = geneKO.iloc[0:-1]\n",
    "    return geneKO\n",
    "\n",
    "def download_seqs(geneList,fileName,batchSize):    \n",
    "    geneList_chunks = [geneList[i * batchSize:(i + 1) * batchSize] for i in range((len(geneList) + batchSize - 1) // batchSize )]\n",
    "    with open(fileName,'w') as fastafile:\n",
    "        for sglist in geneList_chunks:\n",
    "            url =  'http://rest.kegg.jp/get/' + \"+\".join(sglist) + '/aaseq'\n",
    "            fseqs = requests.get(url).text\n",
    "            fastafile.write(fseqs)\n",
    "\n",
    "            \n",
    "def msaSeq2df(seq,indexStart):\n",
    "    idx = []\n",
    "    resid = []\n",
    "    j = indexStart;\n",
    "    for i in seq:\n",
    "        resid.append(i)\n",
    "        if i == '-':\n",
    "            idx.append(np.nan)\n",
    "        else:\n",
    "            idx.append(j)\n",
    "            j = j+1\n",
    "    \n",
    "    return pd.DataFrame({'residue': resid, 'index': idx})\n",
    "    \n",
    "get_resid = lambda x: [x for x in re.findall(r'\\D',x) if len(x)>0][0]\n",
    "get_resid_idx = lambda x: [int(x) for x in re.findall(r'\\d*',x) if len(x)>0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ecogenes():\n",
    "    url = 'http://rest.kegg.jp/link/ko/eco'\n",
    "    r = requests.get(url)\n",
    "    geneKO = pd.DataFrame([x.split('\\t') for x in r.text.split('\\n')],columns=['gene','ko'])\n",
    "    geneKO = geneKO.iloc[0:-1]\n",
    "    return geneKO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "structure = pd.read_csv('../assets/structure/oxidoreductases_residue_annotation.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy = pd.read_csv('../assets/kegg/KEGG_taxonomy_final.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#taxonomy[taxonomy['class'].isin(['Gammaproteobacteria'])]\n",
    "k = 'Gammaproteobacteria'\n",
    "taxons = taxonomy[taxonomy['class'].isin([k])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#genes = ['leuB','pdxB','gdhA','gatD','lgoD','fucO','mdh','icd','paaH','gapA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#s = structure[structure['gene name'].isin(genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = structure.copy()\n",
    "KO = s[s['Coenzyme Bound Structure'] != 'None'].KO.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "76"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(KO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finished downloading :K12524; num seqs: 554\n",
      "finished downloading :K00215; num seqs: 775\n",
      "finished downloading :K00382; num seqs: 1213\n",
      "finished downloading :K00099; num seqs: 756\n",
      "finished downloading :K04073; num seqs: 136\n",
      "finished downloading :K00121; num seqs: 685\n",
      "finished downloading :K00286; num seqs: 821\n",
      "finished downloading :K00077; num seqs: 898\n",
      "finished downloading :K01491; num seqs: 818\n",
      "finished downloading :K00384; num seqs: 746\n",
      "finished downloading :K12972; num seqs: 358\n",
      "finished downloading :K00059; num seqs: 2504\n",
      "finished downloading :K02492; num seqs: 731\n",
      "finished downloading :K00208; num seqs: 534\n",
      "finished downloading :K09472; num seqs: 276\n",
      "finished downloading :K03778; num seqs: 552\n",
      "finished downloading :K13953; num seqs: 354\n",
      "finished downloading :K00027; num seqs: 596\n",
      "finished downloading :K08324; num seqs: 271\n",
      "finished downloading :K16066; num seqs: 408\n",
      "finished downloading :K05887; num seqs: 159\n",
      "finished downloading :K06447; num seqs: 542\n",
      "finished downloading :K00262; num seqs: 535\n",
      "finished downloading :K00134; num seqs: 1380\n",
      "finished downloading :K00036; num seqs: 869\n",
      "finished downloading :K00013; num seqs: 744\n",
      "finished downloading :K00012; num seqs: 763\n",
      "finished downloading :K00033; num seqs: 643\n",
      "finished downloading :K00067; num seqs: 586\n",
      "finished downloading :K02377; num seqs: 266\n",
      "finished downloading :K03473; num seqs: 602\n",
      "finished downloading :K00088; num seqs: 776\n",
      "finished downloading :K00529; num seqs: 172\n",
      "finished downloading :K00135; num seqs: 1108\n",
      "finished downloading :K00058; num seqs: 1061\n",
      "finished downloading :K00266; num seqs: 670\n",
      "finished downloading :K00024; num seqs: 656\n",
      "finished downloading :K00014; num seqs: 1264\n",
      "finished downloading :K02302; num seqs: 745\n",
      "finished downloading :K00133; num seqs: 968\n",
      "finished downloading :K00383; num seqs: 703\n",
      "finished downloading :K00090; num seqs: 434\n",
      "finished downloading :K00057; num seqs: 765\n",
      "finished downloading :K00060; num seqs: 499\n",
      "finished downloading :K00053; num seqs: 752\n",
      "finished downloading :K02472; num seqs: 410\n",
      "finished downloading :K01825; num seqs: 555\n",
      "finished downloading :K00005; num seqs: 420\n",
      "finished downloading :K00145; num seqs: 723\n",
      "finished downloading :K00046; num seqs: 222\n",
      "finished downloading :K00287; num seqs: 783\n",
      "finished downloading :K00052; num seqs: 739\n",
      "finished downloading :K00364; num seqs: 375\n",
      "finished downloading :K00130; num seqs: 611\n",
      "finished downloading :K11752; num seqs: 781\n",
      "finished downloading :K00073; num seqs: 174\n",
      "finished downloading :K00299; num seqs: 415\n",
      "finished downloading :K13821; num seqs: 688\n",
      "finished downloading :K00031; num seqs: 840\n",
      "finished downloading :K04072; num seqs: 428\n",
      "finished downloading :K02618; num seqs: 168\n",
      "finished downloading :K07248; num seqs: 186\n",
      "finished downloading :K00137; num seqs: 316\n",
      "finished downloading :K07246; num seqs: 243\n",
      "finished downloading :K00335; num seqs: 607\n",
      "finished downloading :K14187; num seqs: 489\n",
      "finished downloading :K00380; num seqs: 785\n",
      "finished downloading :K06879; num seqs: 711\n",
      "finished downloading :K00048; num seqs: 204\n",
      "finished downloading :K19265; num seqs: 390\n",
      "finished downloading :K00042; num seqs: 494\n",
      "finished downloading :K08092; num seqs: 198\n",
      "finished downloading :K00528; num seqs: 768\n",
      "finished downloading :K00297; num seqs: 829\n",
      "finished downloading :K00075; num seqs: 783\n",
      "finished downloading :K00040; num seqs: 536\n"
     ]
    }
   ],
   "source": [
    "for ko in KO:\n",
    "    genes = ko2genes(ko)\n",
    "    genes['species'] = genes['gene'].apply(lambda x: x.split(':')[0])\n",
    "    ge = genes[genes.species.isin(taxons.kegg_id.tolist())]\n",
    "    gl = ge.gene.unique().tolist()\n",
    "    download_seqs(gl,'../assets/kegg/orthogroups_class/' + ko + '_Gammaproteobacteria.fasta',10)\n",
    "    print('finished downloading :' + ko + '; num seqs: ' + str(len(gl)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finished with mutiple sequence alignment: K12524; num seqs: \n",
      "finished with mutiple sequence alignment: K00215; num seqs: \n",
      "finished with mutiple sequence alignment: K00382; num seqs: \n",
      "finished with mutiple sequence alignment: K00099; num seqs: \n",
      "finished with mutiple sequence alignment: K04073; num seqs: \n",
      "finished with mutiple sequence alignment: K00121; num seqs: \n",
      "finished with mutiple sequence alignment: K00286; num seqs: \n",
      "finished with mutiple sequence alignment: K00077; num seqs: \n",
      "finished with mutiple sequence alignment: K01491; num seqs: \n",
      "finished with mutiple sequence alignment: K00384; num seqs: \n",
      "finished with mutiple sequence alignment: K12972; num seqs: \n",
      "finished with mutiple sequence alignment: K00059; num seqs: \n",
      "finished with mutiple sequence alignment: K02492; num seqs: \n",
      "finished with mutiple sequence alignment: K00208; num seqs: \n",
      "finished with mutiple sequence alignment: K09472; num seqs: \n",
      "finished with mutiple sequence alignment: K03778; num seqs: \n",
      "finished with mutiple sequence alignment: K13953; num seqs: \n",
      "finished with mutiple sequence alignment: K00027; num seqs: \n",
      "finished with mutiple sequence alignment: K08324; num seqs: \n",
      "finished with mutiple sequence alignment: K16066; num seqs: \n",
      "finished with mutiple sequence alignment: K05887; num seqs: \n",
      "finished with mutiple sequence alignment: K06447; num seqs: \n",
      "finished with mutiple sequence alignment: K00262; num seqs: \n",
      "finished with mutiple sequence alignment: K00134; num seqs: \n",
      "finished with mutiple sequence alignment: K00036; num seqs: \n",
      "finished with mutiple sequence alignment: K00013; num seqs: \n",
      "finished with mutiple sequence alignment: K00012; num seqs: \n",
      "finished with mutiple sequence alignment: K00033; num seqs: \n",
      "finished with mutiple sequence alignment: K00067; num seqs: \n",
      "finished with mutiple sequence alignment: K02377; num seqs: \n",
      "finished with mutiple sequence alignment: K03473; num seqs: \n",
      "finished with mutiple sequence alignment: K00088; num seqs: \n",
      "finished with mutiple sequence alignment: K00529; num seqs: \n",
      "finished with mutiple sequence alignment: K00135; num seqs: \n",
      "finished with mutiple sequence alignment: K00058; num seqs: \n",
      "finished with mutiple sequence alignment: K00266; num seqs: \n",
      "finished with mutiple sequence alignment: K00024; num seqs: \n",
      "finished with mutiple sequence alignment: K00014; num seqs: \n",
      "finished with mutiple sequence alignment: K02302; num seqs: \n",
      "finished with mutiple sequence alignment: K00133; num seqs: \n",
      "finished with mutiple sequence alignment: K00383; num seqs: \n",
      "finished with mutiple sequence alignment: K00090; num seqs: \n",
      "finished with mutiple sequence alignment: K00057; num seqs: \n",
      "finished with mutiple sequence alignment: K00060; num seqs: \n",
      "finished with mutiple sequence alignment: K00053; num seqs: \n",
      "finished with mutiple sequence alignment: K02472; num seqs: \n",
      "finished with mutiple sequence alignment: K01825; num seqs: \n",
      "finished with mutiple sequence alignment: K00005; num seqs: \n",
      "finished with mutiple sequence alignment: K00145; num seqs: \n",
      "finished with mutiple sequence alignment: K00046; num seqs: \n",
      "finished with mutiple sequence alignment: K00287; num seqs: \n",
      "finished with mutiple sequence alignment: K00052; num seqs: \n",
      "finished with mutiple sequence alignment: K00364; num seqs: \n",
      "finished with mutiple sequence alignment: K00130; num seqs: \n",
      "finished with mutiple sequence alignment: K11752; num seqs: \n",
      "finished with mutiple sequence alignment: K00073; num seqs: \n",
      "finished with mutiple sequence alignment: K00299; num seqs: \n",
      "finished with mutiple sequence alignment: K13821; num seqs: \n",
      "finished with mutiple sequence alignment: K00031; num seqs: \n",
      "finished with mutiple sequence alignment: K04072; num seqs: \n",
      "finished with mutiple sequence alignment: K02618; num seqs: \n",
      "finished with mutiple sequence alignment: K07248; num seqs: \n",
      "finished with mutiple sequence alignment: K00137; num seqs: \n",
      "finished with mutiple sequence alignment: K07246; num seqs: \n",
      "finished with mutiple sequence alignment: K00335; num seqs: \n",
      "finished with mutiple sequence alignment: K14187; num seqs: \n",
      "finished with mutiple sequence alignment: K00380; num seqs: \n",
      "finished with mutiple sequence alignment: K06879; num seqs: \n",
      "finished with mutiple sequence alignment: K00048; num seqs: \n",
      "finished with mutiple sequence alignment: K19265; num seqs: \n",
      "finished with mutiple sequence alignment: K00042; num seqs: \n",
      "finished with mutiple sequence alignment: K08092; num seqs: \n",
      "finished with mutiple sequence alignment: K00528; num seqs: \n",
      "finished with mutiple sequence alignment: K00297; num seqs: \n",
      "finished with mutiple sequence alignment: K00075; num seqs: \n",
      "finished with mutiple sequence alignment: K00040; num seqs: \n"
     ]
    }
   ],
   "source": [
    "for ko in KO:\n",
    "    fin = '../assets/kegg/orthogroups_class/' + ko + '_Gammaproteobacteria.fasta'\n",
    "    fout = '../assets/kegg/orthogroups_class/msa/' + ko + '_Gammaproteobacteria.MSA.fasta'\n",
    "    call = '/Users/Joshua.Goldford/opt/miniconda3/bin/muscle -in ' + fin + ' -out ' + fout\n",
    "    subprocess.call(call,shell=True)\n",
    "    print('finished with mutiple sequence alignment: ' + ko + '; num seqs: ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "ko_list =  [];\n",
    "residue_list = []\n",
    "residue_ref_idx = []\n",
    "residue_msa_idx = []\n",
    "count_occured = []\n",
    "count_not_occured = []\n",
    "frac_occured = []\n",
    "\n",
    "for ko in KO:\n",
    "    seq_dict = SeqIO.to_dict(SeqIO.parse('../assets/kegg/orthogroups_class/msa/' + ko + '_Gammaproteobacteria.MSA.fasta','fasta'))\n",
    "    if s[s.KO == ko]['Homolog-KEGG'].tolist()[0] in list(seq_dict):\n",
    "        seq_id = s[s.KO == ko]['Homolog-KEGG'].tolist()[0]\n",
    "        r = s[s.KO == ko].Residues[0].split(';')\n",
    "        r  = [x for x in r if x not in ['Unclear',None]]\n",
    "        if len(r) > 0:\n",
    "            r = pd.DataFrame({'residue': [get_resid(y) for y in r],'index':[get_resid_idx(y) for y in r]})\n",
    "            seq = str(seq_dict[seq_id].seq)\n",
    "            seqdf = msaSeq2df(seq,1)\n",
    "            sdf = seqdf[seqdf['index'].isin(r['index'].tolist())]\n",
    "            for midx,row in sdf.iterrows():\n",
    "                chars = [y.seq[midx] for y in list(seq_dict.values())]\n",
    "                z = pd.DataFrame({'idx':chars}).reset_index().groupby('idx').count()\n",
    "                frac = z.loc[row.residue]['index'] / z['index'].sum()\n",
    "                ko_list.append(ko)\n",
    "                residue_list.append(row.residue)\n",
    "                residue_ref_idx.append(row['index'])\n",
    "                residue_msa_idx.append(midx)\n",
    "                count_occured.append(z.loc[row.residue]['index'])\n",
    "                count_not_occured.append(z['index'].sum() - z.loc[row.residue]['index'])\n",
    "                frac_occured.append(frac)\n",
    "\n",
    "results = pd.DataFrame({'ko':ko_list,'residue':residue_list,'idx': residue_ref_idx, 'midx': residue_msa_idx, 'count': count_occured, 'ncount': count_not_occured, 'frac': frac_occured})    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "z = pd.read_csv('../assets/mutantGeneSet.reactionDirectionUpdateRules.MR4.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>reaction</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>deltaG_prime_m</th>\n",
       "      <th>deltaG_prime_m (error)</th>\n",
       "      <th>wild type coenzyme</th>\n",
       "      <th>mutant coenzyme</th>\n",
       "      <th>max_deltaG (WT)</th>\n",
       "      <th>min_deltaG (WT)</th>\n",
       "      <th>max_deltaG (mutant)</th>\n",
       "      <th>min_deltaG (mutant)</th>\n",
       "      <th>direction (wild type)</th>\n",
       "      <th>direction (mutant)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>5DGLCNR</td>\n",
       "      <td>b4266</td>\n",
       "      <td>idnO</td>\n",
       "      <td>-18.057655</td>\n",
       "      <td>3.247864</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>-5.254373</td>\n",
       "      <td>-50.918449</td>\n",
       "      <td>13.312967</td>\n",
       "      <td>-32.351109</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "      <td>reversible</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ABUTD</td>\n",
       "      <td>b1444</td>\n",
       "      <td>patD</td>\n",
       "      <td>-52.079065</td>\n",
       "      <td>2.011257</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>-37.785612</td>\n",
       "      <td>-83.449688</td>\n",
       "      <td>-19.218272</td>\n",
       "      <td>-64.882348</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>AGPR</td>\n",
       "      <td>b3958</td>\n",
       "      <td>argC</td>\n",
       "      <td>22.212782</td>\n",
       "      <td>1.615149</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>89.321633</td>\n",
       "      <td>20.825519</td>\n",
       "      <td>70.754293</td>\n",
       "      <td>2.258179</td>\n",
       "      <td>irreversible (backward)</td>\n",
       "      <td>irreversible (backward)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>AKGDH</td>\n",
       "      <td>b0116</td>\n",
       "      <td>lpd</td>\n",
       "      <td>-28.188154</td>\n",
       "      <td>3.884974</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>8.937337</td>\n",
       "      <td>-82.390815</td>\n",
       "      <td>27.504677</td>\n",
       "      <td>-63.823475</td>\n",
       "      <td>reversible</td>\n",
       "      <td>reversible</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ALCD19</td>\n",
       "      <td>b0356</td>\n",
       "      <td>frmA</td>\n",
       "      <td>-36.684521</td>\n",
       "      <td>2.006562</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>-5.313898</td>\n",
       "      <td>-50.977974</td>\n",
       "      <td>-23.881238</td>\n",
       "      <td>-69.545314</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>90</td>\n",
       "      <td>UAPGR</td>\n",
       "      <td>b3972</td>\n",
       "      <td>murB</td>\n",
       "      <td>-56.001735</td>\n",
       "      <td>1.299059</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>-43.198453</td>\n",
       "      <td>-88.862529</td>\n",
       "      <td>-24.631113</td>\n",
       "      <td>-70.295189</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>91</td>\n",
       "      <td>UDPGD</td>\n",
       "      <td>b2028</td>\n",
       "      <td>ugd</td>\n",
       "      <td>-26.705204</td>\n",
       "      <td>2.201167</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>-20.950335</td>\n",
       "      <td>-66.614411</td>\n",
       "      <td>16.184345</td>\n",
       "      <td>-29.479731</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "      <td>reversible</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>92</td>\n",
       "      <td>UDPGDC</td>\n",
       "      <td>b2255</td>\n",
       "      <td>arnA</td>\n",
       "      <td>-28.184567</td>\n",
       "      <td>3.850471</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>-25.307132</td>\n",
       "      <td>-93.803246</td>\n",
       "      <td>-6.739792</td>\n",
       "      <td>-75.235906</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>93</td>\n",
       "      <td>URDGLYCD</td>\n",
       "      <td>b0517</td>\n",
       "      <td>allD</td>\n",
       "      <td>-10.528255</td>\n",
       "      <td>5.435085</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>3.765199</td>\n",
       "      <td>-41.898877</td>\n",
       "      <td>22.332538</td>\n",
       "      <td>-23.331538</td>\n",
       "      <td>reversible</td>\n",
       "      <td>reversible</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>94</td>\n",
       "      <td>XAND</td>\n",
       "      <td>b2867</td>\n",
       "      <td>xdhB</td>\n",
       "      <td>-25.183686</td>\n",
       "      <td>4.676524</td>\n",
       "      <td>NAD(H)</td>\n",
       "      <td>NADP(H)</td>\n",
       "      <td>-10.890233</td>\n",
       "      <td>-56.554309</td>\n",
       "      <td>7.677107</td>\n",
       "      <td>-37.986969</td>\n",
       "      <td>irreversible (forward)</td>\n",
       "      <td>reversible</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  reaction   gene gene_name  deltaG_prime_m  \\\n",
       "0            0   5DGLCNR  b4266      idnO      -18.057655   \n",
       "1            1     ABUTD  b1444      patD      -52.079065   \n",
       "2            2      AGPR  b3958      argC       22.212782   \n",
       "3            3     AKGDH  b0116       lpd      -28.188154   \n",
       "4            4    ALCD19  b0356      frmA      -36.684521   \n",
       "..         ...       ...    ...       ...             ...   \n",
       "90          90     UAPGR  b3972      murB      -56.001735   \n",
       "91          91     UDPGD  b2028       ugd      -26.705204   \n",
       "92          92    UDPGDC  b2255      arnA      -28.184567   \n",
       "93          93  URDGLYCD  b0517      allD      -10.528255   \n",
       "94          94      XAND  b2867      xdhB      -25.183686   \n",
       "\n",
       "    deltaG_prime_m (error) wild type coenzyme mutant coenzyme  \\\n",
       "0                 3.247864            NADP(H)          NAD(H)   \n",
       "1                 2.011257             NAD(H)         NADP(H)   \n",
       "2                 1.615149            NADP(H)          NAD(H)   \n",
       "3                 3.884974             NAD(H)         NADP(H)   \n",
       "4                 2.006562             NAD(H)         NADP(H)   \n",
       "..                     ...                ...             ...   \n",
       "90                1.299059            NADP(H)          NAD(H)   \n",
       "91                2.201167             NAD(H)         NADP(H)   \n",
       "92                3.850471             NAD(H)         NADP(H)   \n",
       "93                5.435085             NAD(H)         NADP(H)   \n",
       "94                4.676524             NAD(H)         NADP(H)   \n",
       "\n",
       "    max_deltaG (WT)  min_deltaG (WT)  max_deltaG (mutant)  \\\n",
       "0         -5.254373       -50.918449            13.312967   \n",
       "1        -37.785612       -83.449688           -19.218272   \n",
       "2         89.321633        20.825519            70.754293   \n",
       "3          8.937337       -82.390815            27.504677   \n",
       "4         -5.313898       -50.977974           -23.881238   \n",
       "..              ...              ...                  ...   \n",
       "90       -43.198453       -88.862529           -24.631113   \n",
       "91       -20.950335       -66.614411            16.184345   \n",
       "92       -25.307132       -93.803246            -6.739792   \n",
       "93         3.765199       -41.898877            22.332538   \n",
       "94       -10.890233       -56.554309             7.677107   \n",
       "\n",
       "    min_deltaG (mutant)    direction (wild type)       direction (mutant)  \n",
       "0            -32.351109   irreversible (forward)               reversible  \n",
       "1            -64.882348   irreversible (forward)   irreversible (forward)  \n",
       "2              2.258179  irreversible (backward)  irreversible (backward)  \n",
       "3            -63.823475               reversible               reversible  \n",
       "4            -69.545314   irreversible (forward)   irreversible (forward)  \n",
       "..                  ...                      ...                      ...  \n",
       "90           -70.295189   irreversible (forward)   irreversible (forward)  \n",
       "91           -29.479731   irreversible (forward)               reversible  \n",
       "92           -75.235906   irreversible (forward)   irreversible (forward)  \n",
       "93           -23.331538               reversible               reversible  \n",
       "94           -37.986969   irreversible (forward)               reversible  \n",
       "\n",
       "[95 rows x 14 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "v = results.groupby('ko').mean().sort_values('frac')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def msaSeq2df(seq,indexStart):\n",
    "    idx = []\n",
    "    resid = []\n",
    "    j = indexStart;\n",
    "    for i in seq:\n",
    "        resid.append(i)\n",
    "        if i == '-':\n",
    "            idx.append(np.nan)\n",
    "        else:\n",
    "            idx.append(j)\n",
    "            j = j+1\n",
    "    \n",
    "    return pd.DataFrame({'residue': resid, 'index': idx})\n",
    "    \n",
    "get_resid = lambda x: [x for x in re.findall(r'\\D',x) if len(x)>0][0]\n",
    "get_resid_idx = lambda x: [int(x) for x in re.findall(r'\\d*',x) if len(x)>0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "msa_index[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>idx</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>392</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index\n",
       "idx       \n",
       "-        5\n",
       "D      392"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame({'idx':chars}).reset_index().groupby('idx').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "h = pd.read_csv('../assets/EC_1.X.1.X.Unique_NAD(P)_Coupled.GeneSet.Thermo.csv').set_index('gene').join(structure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "v = v[['frac']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "z = v.join(h.reset_index().set_index('KO')[['deltaG_prime_m']]).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='deltaG_prime_m', ylabel='frac'>"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEHCAYAAACjh0HiAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAa4UlEQVR4nO3df5QdZ33f8ffnyiuvox8YpJVEJNkr5YgayTgyvshuqU/42a4VKiWBqpYpwTSJDhwL0QpOYscchzqQGJKIoFrYlamDSQFZpaRRqbAhJiQOtY3WINvIrm2xKFgCSWuB0Q+y1kr32z/u7PpqdfeHdnfu3LnzeZ2zR3dmnp373blX8515nmeeRxGBmZkVVynrAMzMLFtOBGZmBedEYGZWcE4EZmYF50RgZlZw52UdwLmaPXt2dHZ2Zh2GmVmuPProo89HREe9bblLBJ2dnXR3d2cdhplZrkj6x+G2uWrIzKzgnAjMzArOicDMrOCcCMzMCs6JwMys4FLrNSTpbuBtwOGIuLTOdgGfAlYCPweuj4jvpBWPmRVHpRLsO3KCQ0f7mDuznc5Z0wBGXPcLU8/j5OnTzJp2Pp2zplEqKcs/oaHS7D76WeB24HPDbL8GWJL8XAnckfxrZjZulUpw356DbNy+m77+Cu1tJW6/7nJOnooz1m1as5yp54n1X/ju4LoNb1rCvd0/5Pe6Xk3XsnmFSQapVQ1FxN8DPxmhyGrgc1H1MHChpFemFY+ZFcO+IycGT/gAff0VHt//s7PWbdy+m8f3/+yMdZu/8Sxvu2w+G7fvZt+RE5n9DY2WZRvBfOC5muX9ybqzSFonqVtSd29vb0OCM7N8OnS0b/DkPqASnLWur79CZch0LH39FaTqv4eP9aUdatPIRWNxRGyNiHJElDs66j4hbWYGwNyZ7bS3nXlqmyLOWtfeVmJozU97W4mI6r9zZrSnHWrTyDIRHAAW1iwvSNaZNa1KJejpPc5D33+ent7jVIZeUlrmOmdNY9Oa5YMn/va2Eq9Z8LKz1m1as5zLFrzsjHUb3rSErzx+gE1rlg82JheB0pyqUlIn8JVheg39KrCeaq+hK4HNEbFitH2Wy+XwWEOWhXqNkJvWLC9Uo2JeDPQaOnysjzkzzuwhVG9dtdfQFPpPV3hFi/YakvRoRJTrbksrEUj6IvAGYDZwCPgDoA0gIu5Muo/eDnRR7T76nogY9QzvRGBZ6ek9zsrND55R19zeVmLnhqtZ3DE9w8jMRjdSIkit+2hErB1lewA3pPX+ZpOtXiPkQKOiE4HlWS4ai82aQb1GyKI1KlprciIwG6N6jZBFa1S01pS7iWnMslIqia5l87hkw9VnNDi2WqOiFY8TQQbqjYPik0k+lEpiccd0twlYS3EiaDB3QTSzZuM2ggarNw5K0cY1MbPm4kTQYCN1QTQzy4ITQYO5C6KZNRsnggZzF0QzazZuLG4wd0E0s2bjRJABd0E0ayx32R6ZE4GZtTR32R6d2wjMrKW5y/bonAjMrKW5y/bonAjMrKW5y/bonAjMrKW5y/bo3FhsZi3NXbZH50RgZi3PXbZH5qohM7OCSzURSOqS9LSkvZJurLP9YkkPSHpc0jclLUgzHjMzO1tqiUDSFGALcA2wFFgraemQYn8KfC4iLgNuBf44rXjMzKy+NO8IVgB7I6InIk4C24DVQ8osBb6RvP7bOtutgCqVoKf3OA99/3l6eo9TqUTWIZm1tDQbi+cDz9Us7weuHFLmMeA3gE8Bvw7MkDQrIo5MZiAeZyQ/PByAWeNl3Vj8IeBXJH0X+BXgAHB6aCFJ6yR1S+ru7e09pzcYOLGs3Pwga+96hJWbH+S+PQd9ldmkPByAWeOlmQgOAAtrlhck6wZFxI8i4jci4nLg5mTdC0N3FBFbI6IcEeWOjo5zCsInlnzxcABncjWZNUKaVUO7gCWSFlFNANcC19UWkDQb+ElEVICbgLsnO4iRTizuU9x8BoYDqP3MijocQN6qyVwFm1+p3RFExClgPXA/8BSwPSL2SLpV0qqk2BuApyU9A8wFPjbZcXickXzxcAAvydPdrKtg800R+fqgyuVydHd3j7l83q6q7KUry2YbDqDRV7wPff951t71yFnrt627kqsWz07tfcejp/c4Kzc/eNad3M4NV/vOu0lIejQiyvW2tfwQEx5nJH+acTiALC4o8lRN5irYfMu611BDDJxYrlo8m8Ud01sqCbgxsTGyqKbJUzWZq2DzreXvCFqZq70aJ4sr3jzdzQ4kraHfxWZMWnY2J4IcG+4q9RLXy066rKppmrGarJ48JS07WyGqhlqV+9w3Tp6qabLSylWwrc53BDmWp8bEvPMVr7Uy3xHkmK9SG8tXvNaqfEeQY75KNbPJ4ESQc3lpTDSz5uWqITOzgnMiMDMrOFcNmZmNUauOsOpEYGY2Bq38JL+rhiyXPMaSNVqehgU/V74jsNxp5Ssza15ZjrCadpWU7wgsd1r5ysyaV1YjrDZi0h8nAssdj7FkWcjqSf5GXPi4ashyJ6sxllq1x4iN7bPN6kn+RlRJORFY7mQx9r3bJVrXuXy2WTzJ34gLn5afs9iyldZVdKPnNfacvK2r2T/byboIyWzOYkldwKeAKcBnIuK2IdsvAu4BLkzK3BgRO9OMyRonzavoRl+ZeU7e1tXsn20jqqRSayyWNAXYAlwDLAXWSlo6pNiHge0RcTlwLfDptOKxxmul3j2ek7d15eGzTXsI9DR7Da0A9kZET0ScBLYBq4eUCWBm8vplwI9SjMcarJV693juh9blzzbdqqH5wHM1y/uBK4eU+QjwNUnvB6YBb6m3I0nrgHUAF1100aQHaulopRnUPPdD6/Jnm/1zBGuBz0bEAmAl8JeSzoopIrZGRDkiyh0dHQ0P0san1a60PENZ6yr6Z5vmHcEBYGHN8oJkXa3fAroAIuIhSe3AbOBwinFZg/hKyywf0rwj2AUskbRI0lSqjcE7hpT5IfBmAEmvBtqB3hRjsgYr+pWWWR6klggi4hSwHrgfeIpq76A9km6VtCop9kHgdyQ9BnwRuD7y9mCDmVnOpfocQfJMwM4h626pef0k8Po0YzAzs5Fl3VhsZmYZcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCSzURSOqS9LSkvZJurLP9k5J2Jz/PSHohzXjMzOxsqU1eL2kKsAV4K7Af2CVpRzJhPQAR8Z9qyr8fuDyteMzMrL407whWAHsjoiciTgLbgNUjlF8LfDHFeMzMrI40E8F84Lma5f3JurNIuhhYBHxjmO3rJHVL6u7t7Z30QM3MiqxZGouvBb4UEafrbYyIrRFRjohyR0dHg0MzM2ttaSaCA8DCmuUFybp6rsXVQmZmmUgzEewClkhaJGkq1ZP9jqGFJF0CvBx4KMVYzMxsGKklgog4BawH7geeArZHxB5Jt0paVVP0WmBbRERasZiZ2fBS6z4KEBE7gZ1D1t0yZPkjacZgZmYja5bGYjMzy4gTgZlZwY2aCCT9kaQLa5ZfLumjqUZlZmYNM5Y7gmsi4oWBhYj4KbAytYjMzKyhxpIIpkg6f2BB0gXA+SOUNzOzHBlLr6HPAw9I+otk+T3APemFZGZmjTRqIoiIj0t6HHhzsuoPI+L+dMMyM7NGGdNzBBHxVeCrKcdiZmYZGEuvoask7ZJ0XNJJSaclHW1EcGZmlr6xNBbfTnWugGeBC4DfpjrhjJmZtYAxPVAWEXuBKRFxOiL+AuhKNywzM2uUsbQR/DwZPXS3pE8AP8ZPJJuZtYyxnNDflZRbD5ygOsfA29MMyizPKpWgp/c4D33/eXp6j1OpeGBda24j3hEkE9D/UUS8E+gD/nNDojLLqUoluG/PQTZu301ff4X2thKb1iyna9k8SiVlHZ5ZXSPeESRTR16cVA2Z2Sj2HTkxmAQA+vorbNy+m31HTmQcmdnwxtJG0AN8S9IOqlVDAETEptSiMsupQ0f7BpPAgL7+CoeP9bG4Y3pGUZmNbNg7Akl/mbxcBXwlKTuj5sfMhpg7s532tjP/W7W3lZgzoz2jiMxGN9IdwRWSfhH4IfBfGhSPWa51zprGpjXLz2oj6Jw1LevQzIY1UiK4E3gAWAR016wXEMDiFOMyy6VSSXQtm8clG67m8LE+5sxop3PWNDcUW1MbNhFExGZgs6Q7IuJ9DYzJLNdKJbG4Y7rbBCw3Rn2OYCJJQFKXpKcl7ZV04zBl1kh6UtIeSV8Y73tZetLqF+/+9vnjz6w1jWn00fFInkHYArwV2A/skrQjIp6sKbMEuAl4fUT8VNKctOKx8UmrX3wj+ttXKsG+Iyc4dLSPuTNdRTNRfkaidaU5VMQKYG9E9ETESWAbsHpImd8BtiTTXxIRh1OMx8YhrX7xafe3Hzhprdz8IGvveoSVmx/kvj0HfQU7AX5GonWlmQjmA8/VLO9P1tV6FfAqSd+S9LCkuoPZSVonqVtSd29vb0rhWj0j9Ytvxv0O8Elr8qX9mVl2sh487jxgCfAGqkNd3yXpwqGFImJrRJQjotzR0dHYCAsurX7xafe390lr8vkZidaVZiI4QHWAugELknW19gM7IqI/In4APEM1MViTGOgXP3ACmKx+8Wntd0ARTlqNbrhN+zOz7CginS+PpPOontjfTDUB7AKui4g9NWW6gLUR8W5Js4HvAssj4shw+y2Xy9Hd3T3cZkvBQKPrZPeLT2u/A/tuxobNyWrAzurvS/Mzs3RJejQiynW3pZUIkjdeCfw5MAW4OyI+JulWoDsidkgS8GdUJ7o5DXwsIraNtE8nAhurZjtpTebJu6f3OCs3P3hG9Vd7W4mdG6728wtW10iJILXuowARsRPYOWTdLTWvA9iY/JhNqmZ7sGu4BuxLxnHy9uB2Npmybiw2K4zJbMAuQhuINY4TgVmDTObJ2w23NplSrRoys5dM5sikHtzOJlOqjcVpcGOx5VmzNWBbcWTWWGxmZ2q2BmwzcBuBmVnhORGYmRWcE4GZWcE5EZiZFZwTgZlZwTkRmJkVnBOBmVnBORGYmRWcE4GZWcE5EZiZFZwTgZlZwTkRmJkVnBOBmVnBORGYmRVcqolAUpekpyXtlXRjne3XS+qVtDv5+e004zEzs7OlNh+BpCnAFuCtwH5gl6QdEfHkkKL3RsT6tOIwM7ORpXlHsALYGxE9EXES2AasTvH9zMxsHNJMBPOB52qW9yfrhnq7pMclfUnSwno7krROUrek7t7e3jRiNTMrrKwbi/830BkRlwFfB+6pVygitkZEOSLKHR0dDQ3QzKzVpZkIDgC1V/gLknWDIuJIRLyYLH4GuCLFeMzMrI40E8EuYImkRZKmAtcCO2oLSHplzeIq4KkU4zEzszpS6zUUEackrQfuB6YAd0fEHkm3At0RsQPYIGkVcAr4CXB9WvGYmVl9ioisYzgn5XI5uru7sw7DzCxXJD0aEeV627JuLDYzs4w5EZiZFZwTgZlZwTkRmJkVnBOBmVnBORGYmRWcE4GZWcE5EZiZFZwTgZlZwTkRmJkVnBOBmVnBpTbonJkVS6US7DtygkNH+5g7s53OWdMolZR1WDYGTgRmNmGVSnDfnoNs3L6bvv4K7W0lNq1ZTteyeU4GOeCqITObsH1HTgwmAYC+/gobt+9m35ETGUdmY+FEYGYTduho32ASGNDXX+Hwsb6MIrJz4URgZhM2d2Y77W1nnk7a20rMmdGeUUR2LpwIzGzCOmdNY9Oa5YPJYKCNoHPWtIwjs7FwY7GZTVipJLqWzeOSDVdz+Fgfc2a411CeOBGY2aQolcTijuks7piedSh2jlKtGpLUJelpSXsl3ThCubdLCkl159M0M7P0pJYIJE0BtgDXAEuBtZKW1ik3A/gA8EhasZiZ2fDSvCNYAeyNiJ6IOAlsA1bXKfeHwMcB9zMzM8tAmolgPvBczfL+ZN0gSa8FFkbE/0kxDjMzG0Fm3UcllYBNwAfHUHadpG5J3b29vekHZ2ZWIGkmggPAwprlBcm6ATOAS4FvStoHXAXsqNdgHBFbI6IcEeWOjo4UQzYzK540E8EuYImkRZKmAtcCOwY2RsTPImJ2RHRGRCfwMLAqIrpTjMnMzIZILRFExClgPXA/8BSwPSL2SLpV0qq03tfMzM5Nqg+URcROYOeQdbcMU/YNacZiZmb1eawhM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMiMDMrOCcCM7OCS3XOYrNWUqkE+46c4NDRPubObKdz1jRKJWUdltmEpZoIJHUBnwKmAJ+JiNuGbH8vcANwGjgOrIuIJ9OMyWw8KpXgvj0H2bh9N339FdrbSmxas5yuZfOcDCz3UqsakjQF2AJcAywF1kpaOqTYFyLiNRGxHPgEsCmteMwmYt+RE4NJAKCvv8LG7bvZd+RExpGZTVyabQQrgL0R0RMRJ4FtwOraAhFxtGZxGhApxmM2boeO9g0mgQF9/RUOH+vLKCKzyZNm1dB84Lma5f3AlUMLSboB2AhMBd5Ub0eS1gHrAC666KJJD9RsNHNnttPeVjojGbS3lZgzoz3DqMwmR+a9hiJiS0T8EvB7wIeHKbM1IsoRUe7o6GhsgGZA56xpbFqznPa26n+ZgTaCzlnTMo7MbOLSvCM4ACysWV6QrBvONuCOFOMxG7dSSXQtm8clG67m8LE+5sxwryFrHWkmgl3AEkmLqCaAa4HragtIWhIRzyaLvwo8i1mTKpXE4o7pLO6YnnUoZpMqtUQQEackrQfup9p99O6I2CPpVqA7InYA6yW9BegHfgq8O614zMysvlSfI4iIncDOIetuqXn9gTTf38zMRpd5Y7GZmWXLicDMrOCcCMzMCk4R+XqYV1IvcAJ4PutYmtxsfIxG42M0Oh+jscnDcbo4Iuo+iJW7RAAgqTsiylnH0cx8jEbnYzQ6H6OxyftxctWQmVnBORGYmRVcXhPB1qwDyAEfo9H5GI3Ox2hscn2cctlGYGZmkyevdwRmZjZJnAjMzAouV4lA0vsl/T9JeyR9omb9TZL2Snpa0r/OMsZmIemDkkLS7GRZkjYnx+lxSa/NOsasSPqT5Hv0uKS/knRhzTZ/lxKSupLjsFfSjVnH0wwkLZT0t5KeTM5DH0jWv0LS1yU9m/z78qxjPRe5SQSS3kh1qstfjohlwJ8m65dSHeJ6GdAFfDqZL7mwJC0E/hXww5rV1wBLkp91FHvuh68Dl0bEZcAzwE3g71KtMc45XkSngA9GxFLgKuCG5LjcCDwQEUuAB5Ll3MhNIgDeB9wWES8CRMThZP1qYFtEvBgRPwD2Up0vucg+CfwuZ84BvRr4XFQ9DFwo6ZWZRJexiPhaRJxKFh+mOmkS+LtUa9Q5x4soIn4cEd9JXh8DnqI6Le9q4J6k2D3Ar2US4DjlKRG8Crha0iOS/k7S65L19eZGnt/w6JqEpNXAgYh4bMgmH6f6/gPw1eS1j9FLfCxGIakTuBx4BJgbET9ONh0E5mYV13ikOh/BuZL0N8C8OptuphrrK6jejr0O2C5pcQPDaxqjHKffp1otVGgjHaOI+OukzM1Ub/U/38jYLP8kTQf+J/AfI+Ko9NKUpRERknLVL7+pEkFEvGW4bZLeB3w5qg8+fFtShepAT+c6N3LuDXecJL0GWAQ8lnwxFwDfkbSCgh2nkb5LAJKuB94GvDleepimUMdoFD4Ww5DURjUJfD4ivpysPiTplRHx46TK9fDwe2g+eaoa+l/AGwEkvQqYSnW0vx3AtZLOT+ZHXgJ8O6sgsxQRT0TEnIjojIhOqrfzr42Ig1SP028mvYeuAn5WcytbKJK6qLahrIqIn9ds8nfpJYNzjkuaSrURfUfGMWVO1Sus/wY8FRGbajbt4KWpdt8N/HWjY5uIprojGMXdwN2SvgecBN6dXMntkbQdeJLqbf4NEXE6wzib1U5gJdUG0J8D78k2nEzdDpwPfD25c3o4It6bzKnt7xLDzzmecVjN4PXAu4AnJO1O1v0+cBvV6urfAv4RWJNNeOPjISbMzAouT1VDZmaWAicCM7OCcyIwMys4JwIzs4JzIjAzKzgnAjOzgnMisNyS9BFJHxrLdknXS/rFMe53YzJM9ROSHpO0KXmadKLxvlfSb050P2aTLU8PlJlNxPXA94AfjVRI0nupjtV0VUS8kDxVuxG4AOgf75tLOi8i7hzv75ulyXcEliuSbpb0jKR/AP5Zsu6XJN0n6VFJD0q6ZMjvvAMoA5+XtFvSBZJukbRL0vckbdVLo4bdDLwvIl4AiIiTEXFbRBwdIabjkj6ZTFTygKSOZP03Jf25pG7gA0PuUL6Z/E63pKckvU7Sl5OJTT5as+9/L+nbSdz/daT5EZI4/iSJ428krUjep0fSqnEcbisIJwLLDUlXUB3zZjnV4TIGhiLfCrw/Iq4APgR8uvb3IuJLQDfwzohYHhH/BNweEa+LiEupXu2/TdJMYHoyF8G5mAZ0JxMm/R3wBzXbpkZEOSL+rM7vnYyIMnAn1bFpbgAuBa6XNEvSq4F/B7w+IpYDp4F3jhLHN5I4jgEfBd4K/Dpw6zn+TVYgrhqyPLka+KuBgeIk7QDagX8B/I+aoYDPH8O+3ijpd4FfoDq8+R6qJ/FBqk5V+XHgQuC6iPi/w+yrAtybvP7vwJdrtt17dvFBA4O4PQHsGRgEUFIP1ZE//yVwBbAr+dsuYORRLU8C99Xs88WI6Jf0BNA5wu9ZwTkRWN6VgBeSK+YxkdRO9a6hHBHPSfoI0J6MK39c0qKI+EFE3A/cL+krVEe7HavaAbxOjFDuxeTfSs3rgeXzAAH3RMRNY3zf/pohtQf3GREVSf6/bsNy1ZDlyd8Dv5bU8c8A/g3VkVR/IOnfQnWYYEm/XOd3jwEzktftyb/PqzrByDtqyv0xcIeSCe2TtoN2Rlaq2cd1wD+c0181vAeAd0iak8TyCkkXT9K+zQb5KsFyIyK+I+le4DGqVSS7kk3vpHry/jDQRnV+3aFTdX4WuFPSPwH/HLiLai+igzX7AbiDal37I5JeBI4D3wK+O0JoJ4AVyfsfplqvP2ER8WSyz69JKlHttXQD1WGOzSaNh6E2myBJxyNietZxmI2Xq4bMzArOVUNmYyTpEc7ukfSuRt8NjBDHE42Mw1qHq4bMzArOVUNmZgXnRGBmVnBOBGZmBedEYGZWcP8foWhbwxb7kMAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.scatterplot(x='deltaG_prime_m',y='frac',data=z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.read_csv('../assets/mutantGeneSet.reactionDirectionUpdateRules.MR4.csv')\n",
    "fba_results = pd.read_hdf('../scc/relativeGrowthRates.StoichAndThermo.MR4.hdf','df')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoff = 1e-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = (fba_results < cutoff).sum() > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "essential = g[g].index.get_level_values(0).tolist()\n",
    "nonessential = g[~g].index.get_level_values(0).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "egenes = ecogenes()\n",
    "egenes['gene'] = egenes['gene'].apply(lambda x: x.split(':')[1])\n",
    "egenes['ko'] = egenes['ko'].apply(lambda x: x.split(':')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "essential_ko = egenes[egenes.gene.isin(essential)].ko.tolist()\n",
    "nonessential_ko = egenes[egenes.gene.isin(nonessential)].ko.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ko</th>\n",
       "      <th>residue</th>\n",
       "      <th>idx</th>\n",
       "      <th>midx</th>\n",
       "      <th>count</th>\n",
       "      <th>ncount</th>\n",
       "      <th>frac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>K00134</td>\n",
       "      <td>D</td>\n",
       "      <td>34.0</td>\n",
       "      <td>232</td>\n",
       "      <td>1147</td>\n",
       "      <td>233</td>\n",
       "      <td>0.831159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>K03473</td>\n",
       "      <td>D</td>\n",
       "      <td>146.0</td>\n",
       "      <td>223</td>\n",
       "      <td>602</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>K00048</td>\n",
       "      <td>K</td>\n",
       "      <td>39.0</td>\n",
       "      <td>39</td>\n",
       "      <td>168</td>\n",
       "      <td>36</td>\n",
       "      <td>0.823529</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ko residue    idx  midx  count  ncount      frac\n",
       "23  K00134       D   34.0   232   1147     233  0.831159\n",
       "28  K03473       D  146.0   223    602       0  1.000000\n",
       "46  K00048       K   39.0    39    168      36  0.823529"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[results.ko.isin(essential_ko)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ko</th>\n",
       "      <th>residue</th>\n",
       "      <th>idx</th>\n",
       "      <th>midx</th>\n",
       "      <th>count</th>\n",
       "      <th>ncount</th>\n",
       "      <th>frac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K00215</td>\n",
       "      <td>E</td>\n",
       "      <td>38.0</td>\n",
       "      <td>86</td>\n",
       "      <td>461</td>\n",
       "      <td>314</td>\n",
       "      <td>0.594839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K00382</td>\n",
       "      <td>V</td>\n",
       "      <td>201.0</td>\n",
       "      <td>723</td>\n",
       "      <td>727</td>\n",
       "      <td>486</td>\n",
       "      <td>0.599340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K00099</td>\n",
       "      <td>V</td>\n",
       "      <td>20.0</td>\n",
       "      <td>49</td>\n",
       "      <td>695</td>\n",
       "      <td>61</td>\n",
       "      <td>0.919312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K00099</td>\n",
       "      <td>L</td>\n",
       "      <td>47.0</td>\n",
       "      <td>79</td>\n",
       "      <td>372</td>\n",
       "      <td>384</td>\n",
       "      <td>0.492063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K00099</td>\n",
       "      <td>E</td>\n",
       "      <td>48.0</td>\n",
       "      <td>80</td>\n",
       "      <td>359</td>\n",
       "      <td>397</td>\n",
       "      <td>0.474868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>K00286</td>\n",
       "      <td>N</td>\n",
       "      <td>36.0</td>\n",
       "      <td>102</td>\n",
       "      <td>112</td>\n",
       "      <td>709</td>\n",
       "      <td>0.136419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>K00286</td>\n",
       "      <td>R</td>\n",
       "      <td>37.0</td>\n",
       "      <td>103</td>\n",
       "      <td>83</td>\n",
       "      <td>738</td>\n",
       "      <td>0.101096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>K00286</td>\n",
       "      <td>S</td>\n",
       "      <td>38.0</td>\n",
       "      <td>104</td>\n",
       "      <td>452</td>\n",
       "      <td>369</td>\n",
       "      <td>0.550548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>K00286</td>\n",
       "      <td>K</td>\n",
       "      <td>41.0</td>\n",
       "      <td>107</td>\n",
       "      <td>349</td>\n",
       "      <td>472</td>\n",
       "      <td>0.425091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>K00384</td>\n",
       "      <td>H</td>\n",
       "      <td>176.0</td>\n",
       "      <td>440</td>\n",
       "      <td>711</td>\n",
       "      <td>35</td>\n",
       "      <td>0.953083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>K00384</td>\n",
       "      <td>R</td>\n",
       "      <td>177.0</td>\n",
       "      <td>441</td>\n",
       "      <td>731</td>\n",
       "      <td>15</td>\n",
       "      <td>0.979893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>K00384</td>\n",
       "      <td>F</td>\n",
       "      <td>181.0</td>\n",
       "      <td>447</td>\n",
       "      <td>535</td>\n",
       "      <td>211</td>\n",
       "      <td>0.717158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>K16066</td>\n",
       "      <td>R</td>\n",
       "      <td>32.0</td>\n",
       "      <td>115</td>\n",
       "      <td>408</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>K16066</td>\n",
       "      <td>R</td>\n",
       "      <td>33.0</td>\n",
       "      <td>116</td>\n",
       "      <td>408</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>K06447</td>\n",
       "      <td>E</td>\n",
       "      <td>173.0</td>\n",
       "      <td>189</td>\n",
       "      <td>519</td>\n",
       "      <td>23</td>\n",
       "      <td>0.957565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>K00013</td>\n",
       "      <td>D</td>\n",
       "      <td>59.0</td>\n",
       "      <td>95</td>\n",
       "      <td>733</td>\n",
       "      <td>11</td>\n",
       "      <td>0.985215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>K00012</td>\n",
       "      <td>D</td>\n",
       "      <td>29.0</td>\n",
       "      <td>56</td>\n",
       "      <td>682</td>\n",
       "      <td>81</td>\n",
       "      <td>0.893840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>K00067</td>\n",
       "      <td>D</td>\n",
       "      <td>30.0</td>\n",
       "      <td>138</td>\n",
       "      <td>252</td>\n",
       "      <td>334</td>\n",
       "      <td>0.430034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>K02377</td>\n",
       "      <td>R</td>\n",
       "      <td>36.0</td>\n",
       "      <td>62</td>\n",
       "      <td>234</td>\n",
       "      <td>32</td>\n",
       "      <td>0.879699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>K00024</td>\n",
       "      <td>D</td>\n",
       "      <td>34.0</td>\n",
       "      <td>79</td>\n",
       "      <td>538</td>\n",
       "      <td>118</td>\n",
       "      <td>0.820122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>K00133</td>\n",
       "      <td>R</td>\n",
       "      <td>10.0</td>\n",
       "      <td>48</td>\n",
       "      <td>686</td>\n",
       "      <td>282</td>\n",
       "      <td>0.708678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>K00133</td>\n",
       "      <td>T</td>\n",
       "      <td>36.0</td>\n",
       "      <td>100</td>\n",
       "      <td>223</td>\n",
       "      <td>745</td>\n",
       "      <td>0.230372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>K00133</td>\n",
       "      <td>S</td>\n",
       "      <td>38.0</td>\n",
       "      <td>102</td>\n",
       "      <td>778</td>\n",
       "      <td>190</td>\n",
       "      <td>0.803719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>K00383</td>\n",
       "      <td>R</td>\n",
       "      <td>198.0</td>\n",
       "      <td>246</td>\n",
       "      <td>672</td>\n",
       "      <td>31</td>\n",
       "      <td>0.955903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>K00383</td>\n",
       "      <td>K</td>\n",
       "      <td>199.0</td>\n",
       "      <td>247</td>\n",
       "      <td>412</td>\n",
       "      <td>291</td>\n",
       "      <td>0.586060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>K00383</td>\n",
       "      <td>R</td>\n",
       "      <td>204.0</td>\n",
       "      <td>252</td>\n",
       "      <td>613</td>\n",
       "      <td>90</td>\n",
       "      <td>0.871977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>K00053</td>\n",
       "      <td>R</td>\n",
       "      <td>68.0</td>\n",
       "      <td>129</td>\n",
       "      <td>720</td>\n",
       "      <td>32</td>\n",
       "      <td>0.957447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>K00053</td>\n",
       "      <td>R</td>\n",
       "      <td>76.0</td>\n",
       "      <td>137</td>\n",
       "      <td>482</td>\n",
       "      <td>270</td>\n",
       "      <td>0.640957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>K00053</td>\n",
       "      <td>S</td>\n",
       "      <td>78.0</td>\n",
       "      <td>139</td>\n",
       "      <td>629</td>\n",
       "      <td>123</td>\n",
       "      <td>0.836436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>K00031</td>\n",
       "      <td>Y</td>\n",
       "      <td>345.0</td>\n",
       "      <td>797</td>\n",
       "      <td>480</td>\n",
       "      <td>360</td>\n",
       "      <td>0.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>K00031</td>\n",
       "      <td>Y</td>\n",
       "      <td>391.0</td>\n",
       "      <td>887</td>\n",
       "      <td>492</td>\n",
       "      <td>348</td>\n",
       "      <td>0.585714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>K08092</td>\n",
       "      <td>E</td>\n",
       "      <td>306.0</td>\n",
       "      <td>347</td>\n",
       "      <td>173</td>\n",
       "      <td>25</td>\n",
       "      <td>0.873737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>K00297</td>\n",
       "      <td>T</td>\n",
       "      <td>59.0</td>\n",
       "      <td>113</td>\n",
       "      <td>718</td>\n",
       "      <td>111</td>\n",
       "      <td>0.866104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>K00075</td>\n",
       "      <td>Y</td>\n",
       "      <td>196.0</td>\n",
       "      <td>321</td>\n",
       "      <td>697</td>\n",
       "      <td>86</td>\n",
       "      <td>0.890166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>K00075</td>\n",
       "      <td>K</td>\n",
       "      <td>272.0</td>\n",
       "      <td>508</td>\n",
       "      <td>697</td>\n",
       "      <td>86</td>\n",
       "      <td>0.890166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>K00040</td>\n",
       "      <td>L</td>\n",
       "      <td>59.0</td>\n",
       "      <td>175</td>\n",
       "      <td>501</td>\n",
       "      <td>35</td>\n",
       "      <td>0.934701</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ko residue    idx  midx  count  ncount      frac\n",
       "0   K00215       E   38.0    86    461     314  0.594839\n",
       "1   K00382       V  201.0   723    727     486  0.599340\n",
       "2   K00099       V   20.0    49    695      61  0.919312\n",
       "3   K00099       L   47.0    79    372     384  0.492063\n",
       "4   K00099       E   48.0    80    359     397  0.474868\n",
       "5   K00286       N   36.0   102    112     709  0.136419\n",
       "6   K00286       R   37.0   103     83     738  0.101096\n",
       "7   K00286       S   38.0   104    452     369  0.550548\n",
       "8   K00286       K   41.0   107    349     472  0.425091\n",
       "10  K00384       H  176.0   440    711      35  0.953083\n",
       "11  K00384       R  177.0   441    731      15  0.979893\n",
       "12  K00384       F  181.0   447    535     211  0.717158\n",
       "20  K16066       R   32.0   115    408       0  1.000000\n",
       "21  K16066       R   33.0   116    408       0  1.000000\n",
       "22  K06447       E  173.0   189    519      23  0.957565\n",
       "24  K00013       D   59.0    95    733      11  0.985215\n",
       "25  K00012       D   29.0    56    682      81  0.893840\n",
       "26  K00067       D   30.0   138    252     334  0.430034\n",
       "27  K02377       R   36.0    62    234      32  0.879699\n",
       "31  K00024       D   34.0    79    538     118  0.820122\n",
       "35  K00133       R   10.0    48    686     282  0.708678\n",
       "36  K00133       T   36.0   100    223     745  0.230372\n",
       "37  K00133       S   38.0   102    778     190  0.803719\n",
       "38  K00383       R  198.0   246    672      31  0.955903\n",
       "39  K00383       K  199.0   247    412     291  0.586060\n",
       "40  K00383       R  204.0   252    613      90  0.871977\n",
       "41  K00053       R   68.0   129    720      32  0.957447\n",
       "42  K00053       R   76.0   137    482     270  0.640957\n",
       "43  K00053       S   78.0   139    629     123  0.836436\n",
       "44  K00031       Y  345.0   797    480     360  0.571429\n",
       "45  K00031       Y  391.0   887    492     348  0.585714\n",
       "49  K08092       E  306.0   347    173      25  0.873737\n",
       "53  K00297       T   59.0   113    718     111  0.866104\n",
       "54  K00075       Y  196.0   321    697      86  0.890166\n",
       "55  K00075       K  272.0   508    697      86  0.890166\n",
       "56  K00040       L   59.0   175    501      35  0.934701"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[results.ko.isin(nonessential_ko)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8 (Coenzyme Paper)",
   "language": "python",
   "name": "coenzyme_paper"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
